File: 2022 maj/c++/3.cpp
#include <iostream>
#include <fstream>
#include <algorithm>
#include <vector>
using namespace std;

int a[200];

bool trojka(int x, int y, int z)
{
    if (y % x != 0)
        return false;
    if (z % y != 0)
        return false;

    return true;
}

int main()
{
    ifstream in("liczby.txt");
    ofstream out("trojki.txt");

    for (int i = 0; i < 200; i++)
        in >> a[i];
    sort(a, a + 200);

    vector <pair <int, int>> three;
    for (int x = 0; x < 200; x++)
    {
        for (int y = x + 1; y < 200; y++)
        {
            for (int z = y + 1; z < 200; z++)
            {
                if (trojka(a[x], a[y], a[z]))
                {
                    out << a[x] << ' ' << a[y] << ' ' << a[z] << '\n';
                    three.push_back({a[x], a[z]});
                }
            }
        }
    }
    cout << "a) " << three.size() << '\n';

    int ans = 0;
    sort(three.begin(), three.end());
    for (int i = 0; i < three.size() - 1; i++)
    {
        for (int j = i + 1; j < three.size(); j++)
        {
            if (three[i].second == three[j].first)
                ans++;
        }
    }
    cout << "b) " << ans << '\n';
}

File: 2022 maj/python/3.py
with open("liczby.txt", "r") as f:
    file = list(map(int, f.read().strip().split('\n')))

def good(x, y, z):
    if y % x != 0:
        return False
    if z % y != 0:
        return False
    return True

file.sort() # każda dobra trójka (x, y, z) spełnia zależność x < y < z
three = []
with open("trojki.txt", "w") as f:
    for i in range(len(file)):
        for j in range(i + 1, len(file)):
            for k in range(j + 1, len(file)):
                if good(file[i], file[j], file[k]):
                    three.append((file[i], file[k]))
                    f.write(str(file[i]) + ' ' + str(file[j]) + ' ' + str(file[k]) + '\n')
print("a)", len(three))

# dobrą piątkę można stworzyć z dwóch dobrych trójek
# jeżeli ostatnia liczba pierwszej trójki jest taka sama jak pierwsza liczba drugiej
three.sort()
ans = 0
for i in range(len(three)):
    for j in range(i + 1, len(three)):
        if three[i][1] == three[j][0]:
            ans += 1
print("b)", ans)